src/query/ts/m3db/encoded_unconsolidated_step_iterator.go

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/query/block"
+	xts "github.com/m3db/m3/src/query/ts"
+)
+
+type encodedStepIterUnconsolidated struct {
+	mu          sync.RWMutex
+	lastBlock   bool
+	exhausted   bool
+	started     bool
+	meta        block.Metadata
+	validIters  []bool
+	seriesMeta  []block.SeriesMeta
+	seriesIters []encoding.SeriesIterator
+}
+
+func (b *encodedBlock) stepIterUnconsolidated() block.UnconsolidatedStepIter {
+	return &encodedStepIterUnconsolidated{
+		meta:        b.meta,
+		seriesMeta:  b.seriesMetas,
+		seriesIters: b.seriesBlockIterators,
+		lastBlock:   b.lastBlock,
+	}
+}
+
+type encodedUnconsolidatedStep struct {
+	time   time.Time
+	values []xts.Datapoints
+}
+
+func (s *encodedUnconsolidatedStep) Time() time.Time          { return s.time }
+func (s *encodedUnconsolidatedStep) Values() []xts.Datapoints { return s.values }
+
+func (it *encodedStepIterUnconsolidated) CurrentUnconsolidated() (
+	block.UnconsolidatedStep,
+	error,
+) {
+	it.mu.RLock()
```

Comment:
Does this actually need to be thread safe? Iterators often times are not (with no ill effects given proper usage). Are the other iterators thread safe?

Reply:
I think it's probably not necessary at the moment, but the next few commits for the block iterator are likely to require this to be threadsafe since there may be async communications between them

Reply:
Hopefully this should be addressed a bit in the design doc for the next phase